gax-grpc/src/main/java/com/google/api/gax/grpc/CallOptionsUtil.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2017, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.grpc;
+
+import com.google.common.collect.ImmutableMap;
+import io.grpc.CallOptions;
+import io.grpc.Metadata;
+import io.grpc.Metadata.Key;
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * A utility class that provides helper functions to work with custom call options. Note, the custom
+ * call options are an experimental feature. On the other hand they were introduced in may 2016 and
+ * seem relatively stable now.
+ */
+class CallOptionsUtil {
+  // this is a call option name, not a header name, it is not transferred over the wire
+  private static final CallOptions.Key<Map<Key<String>, String>> DYNAMIC_HEADERS_CALL_OPTION_KEY =
+      CallOptions.Key.of("x-goog-dynamic-headers", Collections.<Key<String>, String>emptyMap());
```

Comment:
Are the keys of CallOptions supposed to follow similar naming patterns as headers? (Your key name here implies it's a header, even though in your comment you say it's not)

Reply:
No, they are not supposed to follow any patterns. They technically don't even need to have a reasonable unique name, since CallOptions.Key uses reference equality. Any naming suggestions are welcome.

Reply:
I think that snake_case is a good neutral form. Maybe "gax_dynamic_headers".